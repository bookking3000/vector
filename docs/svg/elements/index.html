<!DOCTYPE html><html><head><title>https://svg-tutorial.github.io/elements/</title><link rel="canonical" href="https://svg-tutorial.github.io/"/><meta name="robots" content="noindex"><meta charset="utf-8" /><meta http-equiv="refresh" content="0; url=https://svg-tutorial.github.io/elements/" /></head></html>