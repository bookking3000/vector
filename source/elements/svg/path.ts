import Shape from './shape.js';
import Element from './element.js';
import Circle from './circle.js';
import Ellipse from './ellipse.js';
import Line from './line.js';
import Rectangle from './rectangle.js';
import Group from './group.js';
import SVG from './svg.js';
import Definitions from './definitions.js';

/**
* A path element allows for the creation of complicated shapes and curves.
*/
export default class Path extends Shape {

  // make the type of the root to be more specific
  // TODO: this crazy type conversion is because typescript is complaining that
  // the SVGPathElement does not have the right properties to be considered a
  // SVGGeometryElement, but the specification says that the path elements the
  // geometric shape... so what gives?
  root: any|SVGGeometryElement|SVGPathElement;

  /**
  * Construct a new path element with a string of commands.
  */
  constructor( d:string ) {
    // TODO: see comment above the type of the root
    let path = document.createElementNS( 'http://www.w3.org/2000/svg', 'path') as any;
    path.setAttribute('d', d);
    super(path);
  }

  /**
  * Returns the d attribute
  */
  get d():string {
    return this.root.getAttribute('d');
  }

  /**
  * Sets the d attribute
  */
  set d( d:string ) {
    this.root.setAttribute('d', d);
  }

  attatchArrow( defs:Definitions, start:boolean = true ) {

    if ( defs === undefined ) {
      throw new Error(`Undefined definitions ${this}: ${this.id}`);
    }

    defs.root.innerHTML = `<marker id="arrow" refX="10" refY="5" markerWidth="10" markerHeight="10" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" style="fill:#404040; stroke:none;"></path></marker>`;

    if( start ) {
      this.setAttribute('marker-start', `url(#arrow)`);
    } else {
      this.setAttribute('marker-end', `url(#arrow)`);
    }
  }

  /**
  * Returns the path representation of the provided shape.
  */
  static getPath( shape:Shape ) : Path {

    if ( shape instanceof Circle ) {
      return new Path(`M ${shape.cx + shape.r} ${shape.cy} A ${shape.r} ${shape.r} 0 0 0 ${shape.cx - shape.r} ${shape.cy} A ${shape.r} ${shape.r} 0 0 0 ${shape.cx + shape.r} ${shape.cy}`);
    } else if ( shape instanceof Ellipse ) {
      throw Error('Not Implemented');
    } else if ( shape instanceof Line ) {
      throw Error('Not Implemented');
    } else if ( shape instanceof Path ) {
      throw Error('Not Implemented');
    } else if ( shape instanceof Rectangle ) {
      throw Error('Not Implemented');
    } else {
      throw Error('Not Implemented');
    }
  }
}
