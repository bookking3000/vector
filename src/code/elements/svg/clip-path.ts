import Circle from './circle'
import Ellipse from './ellipse'
import Line from './line'
import Path from './path'
import Polygon from './polygon'
import Rectangle from './rectangle'

import Element from './element'
import { Descriptive, Shape } from './content-model'
import Description from './description'
import MetaData from './meta-data'
import Title from './title'

export default class ClipPath extends Element implements Descriptive, Shape {

  root:SVGClipPathElement;

  constructor() {
    let clipPath = document.createElementNS( 'http://www.w3.org/2000/svg', 'clipPath') as SVGClipPathElement;
    super(clipPath);
  }

  circle(cx: number, cy: number, r: number): Circle {
    return this.appendChild(new Circle(cx, cy, r));
  }
  ellipse(cx: number, cy: number, rx: number, ry: number): Ellipse {
    return this.appendChild(new Ellipse(cx, cy, rx, ry));
  }
  line(x1: number, y1: number, x2: number, y2: number): Line {
    return this.appendChild(new Line(x1, y1, x2, y2));
  }
  path(d: string): Path {
    return this.appendChild(new Path(d));
  }
  polygon(points: string): Polygon {
    return this.appendChild(new Polygon(points));
  }
  rectangle(x: number, y: number, width: number, height: number): Rectangle {
    return this.appendChild(new Rectangle(x, y, width, height));
  }


  description(): Description {
    return this.appendChild(new Description());
  }
  metadata(): MetaData {
    return this.appendChild(new MetaData());
  }
  title(): Title {
    return this.appendChild(new Title());
  }
}
