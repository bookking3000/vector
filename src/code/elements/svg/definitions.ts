import { Marker } from '../../index';
import Element from './element'

export default class Definitions extends Element {
  constructor() {
    let defs = document.createElementNS( 'http://www.w3.org/2000/svg', 'defs');
    super(defs);
  }

  marker( refX:number, refY:number, width:number, height:number ) {
    return this.appendChild(new Marker(refX, refY, width, height));
  }
}
