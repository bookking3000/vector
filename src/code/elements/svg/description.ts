import Element from './element'

export default class Description extends Element {
  constructor() {
    let desc = document.createElementNS( 'http://www.w3.org/2000/svg', 'desc');
    super(desc);
  }
}
