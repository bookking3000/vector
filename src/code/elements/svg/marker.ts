import Element, { CoreAttributes } from './element'
import { Structural, Shape } from './content-model'

import Circle from './circle'
import ClipPath from './clip-path'
import Defs from './definitions'
import Description from './definitions'
import Ellipse from './ellipse'
import Group from './group'
import Line from './line'
import MetaData from './meta-data'
import Path from './path'
import Polygon from './polygon'
import Rectangle from './rectangle'
import Symbol from './symbol'
import SVG from './svg'
import Text from './text'
import Title from './title'
import Use from './use'

type MarkerAttributes = 'viewBox' | 'preserveAspectRatio' | 'refX' | 'refY' | 'markerUnits' | 'markerWidth' | 'markerHeight' | 'orient';

/**
* A marker is a shape that can be repeatably drawn on a shape.
*/
export default class Marker extends Element implements Shape, Structural {

  // make the type of the root to be more specific
  root: SVGMarkerElement;

  /**
  * Constructs a rectangle element at the position (x,y)
  */
  constructor( refX:number, refY:number, width:number, height:number ) {
    let element = document.createElementNS( 'http://www.w3.org/2000/svg', 'marker') as SVGMarkerElement;
    element.setAttributeNS(null, 'refX', refX.toString());
    element.setAttributeNS(null, 'refY', refY.toString());
    element.setAttributeNS(null, 'markerWidth', width.toString());
    element.setAttributeNS(null, 'markerHeight', height.toString());
    super(element);
  }

  // comment inherited from base class
  setAttribute(name: MarkerAttributes | CoreAttributes, value: string): Element {
    this.root.setAttribute(name,value);
    return this;
  }

  // comment inherited from base class
  getAttribute(name: MarkerAttributes | CoreAttributes): string {
    return this.root.getAttribute(name);
  }

  get viewBox():string {
    return this.getAttribute('viewBox');
  }

  set viewBox( value:string ) {
    this.setAttribute('viewBox', value);
  }

  get refX():number {
    return this.root.refX.baseVal.value;
  }

  set refX(value:number) {
    this.root.refX.baseVal.value = value;
  }

  get refY():number {
    return this.root.refY.baseVal.value;
  }

  set refY(value:number) {
    this.root.refY.baseVal.value = value;
  }

  get width():number{
    return this.root.markerWidth.baseVal.value;
  }

  set width(value:number){
    this.root.markerWidth.baseVal.value = value;
  }

  get height():number{
    return this.root.markerHeight.baseVal.value;
  }

  set height(value:number){
    this.root.markerHeight.baseVal.value = value;
  }

  // Descriptive methods

  description(): Description {
    return this.appendChild(new Description());
  }
  metadata(): MetaData {
    return this.appendChild(new MetaData());
  }
  title(): Title {
    return this.appendChild(new Title());
  }

  // Structural methods

  defs(): Defs {
    return this.appendChild(new Defs());
  }

  group(): Group {
    return this.appendChild(new Group());
  }

  svg(): SVG {
    return this.appendChild(new SVG());
  }

  symbol(): Symbol{
    return this.appendChild(new Symbol());
  }

  use(x:number, y:number, width:number, height:number): Use {
    return this.appendChild(new Use(x, y, width, height));
  }

  // Shape methods

  circle(cx: number, cy: number, r: number): Circle {
    return this.appendChild(new Circle(cx, cy, r));
  }
  ellipse( cx:number, cy:number, rx:number, ry:number): Ellipse {
    return this.appendChild(new Ellipse(cx, cy, rx, ry));
  }
  line( x1:number, y1:number, x2:number, y2:number ): Line {
    return this.appendChild(new Line(x1, y1, x2, y2));
  }
  path( d:string ): Path {
    return this.appendChild(new Path(d));
  }
  polygon( points:string ): Polygon {
    return this.appendChild(new Polygon(points));
  }
  rectangle( x:number, y:number, width:number, height:number ): Rectangle {
    return this.appendChild(new Rectangle(x, y, width, height));
  }

  // other methods

  text(x:number, y:number, str:string ){
    return this.appendChild(new Text(x, y, str));
  }

}
