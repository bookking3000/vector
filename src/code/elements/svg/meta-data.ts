import Element from './element'

export default class MetaData extends Element {
  constructor() {
    let metadata = document.createElementNS( 'http://www.w3.org/2000/svg', 'metadata');
    super(metadata);
  }
}
