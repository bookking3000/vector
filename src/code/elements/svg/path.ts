import Shape from './shape'
import Element from './element'
import Circle from './circle'
import Ellipse from './ellipse'
import Line from './line'
import Rectangle from './rectangle'
import Group from './group'
import SVG from './svg'
import Definitions from './definitions'

/**
* A path element allows for the creation of complicated shapes and curves.
*/
export default class Path extends Shape {

  // make the type of the root to be more specific
  // TODO: this crazy type conversion is because typescript is complaining that
  // the SVGPathElement does not have the right properties to be considered a
  // SVGGeometryElement, but the specification says that the path elements the
  // geometric shape... so what gives?
  root: any|SVGGeometryElement|SVGPathElement;

  /**
  * Construct a new path element with a string of commands.
  */
  constructor( d:string ) {
    // TODO: see comment above the type of the root
    let path = document.createElementNS( 'http://www.w3.org/2000/svg', 'path') as any;
    path.setAttribute('d', d);
    super(path);
  }

  /**
  * Returns the d attribute
  */
  get d():string {
    return this.root.getAttribute('d');
  }

  /**
  * Sets the d attribute
  */
  set d( d:string ) {
    this.root.setAttribute('d', d);
  }

  attatchArrow( defs:Definitions, start:boolean = true ) {

    if ( defs === undefined ) {
      throw new Error(`Undefined definitions ${this}: ${this.id}`);
    }

    defs.root.innerHTML = `<marker id="arrow" refX="10" refY="5" markerWidth="10" markerHeight="10" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" style="fill:#404040; stroke:none;"></path></marker>`;

    if( start ) {
      this.setAttribute('marker-start', `url(#arrow)`);
    } else {
      this.setAttribute('marker-end', `url(#arrow)`);
    }
  }

  /**
  * Returns the path representation of the provided shape.
  */
  static getPath( shape:Shape ) : Path {

    if ( shape instanceof Circle ) {
      return new Path(`M ${shape.cx + shape.r} ${shape.cy} A ${shape.r} ${shape.r} 0 0 0 ${shape.cx - shape.r} ${shape.cy} A ${shape.r} ${shape.r} 0 0 0 ${shape.cx + shape.r} ${shape.cy}`);
    } else if ( shape instanceof Ellipse ) {
      throw Error('Not Implemented');
    } else if ( shape instanceof Line ) {
      throw Error('Not Implemented');
    } else if ( shape instanceof Path ) {
      throw Error('Not Implemented');
    } else if ( shape instanceof Rectangle ) {
      throw Error('Not Implemented');
    } else {
      throw Error('Not Implemented');
    }
  }
}
