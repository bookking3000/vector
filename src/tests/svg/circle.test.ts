import { SVG } from '../..';
import { Circle } from '../../elements/svg/circle'

function createSVG() {
  let container = document.createElement('div');
  let svg = container.appendChild(new SVG().root);
}

beforeEach(() => {
  createSVG();
});

it('renders correctly', () => {
  const circle = new Circle(1, 2, 3)
  expect(circle.root.outerHTML).toMatchSnapshot();
});