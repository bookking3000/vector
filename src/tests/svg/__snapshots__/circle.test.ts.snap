// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `"<circle cx=\\"1\\" cy=\\"2\\" r=\\"3\\" id=\\"circle-1\\" class=\\"circle\\"></circle>"`;
